changelog/dmd.source-archive.dd

```diff
@@ -0,0 +1,95 @@
+sub-modules. The std directory wouldn't even need to exist.
+
+The file format of the .sar file is very similar to that of object file libraries
+and various other schemes. It does not adhere to those other schemes due to their
+variances from platform to platform, all the code needed to support things that
+are unneeded for .sar files, and special consideration for D's needs. The format
+is meant to be friendly for memory-mapped file access, and does not have alignhment
+issues.
+
+A .sar file consists of the following sections:
+
+1. a header, to identify it as a .sar file with a magic number and a version
+
+2. a table of contents, one entry per source file. The entries consist of
+an offset/length to the filename string, and an offset/length to the file
+contents
+
+3. the filename strings, each string has a terminating 0
+
+4. the file contents, each file has four 0 bytes appended, as the lexer wants
+that as a sentinel
+
+To create a .sar file, such as one for Phobos' std:
+
+dmd -srcar=/home/duser/dmd/src/phobos/std
+
+and the file:
+
+/home/duser/dmd/src/phobos/std.sar
+
+will be created and filled with all sub-files with one of the extensions ".di", ".d",
+".c", or ".i". The filenames in the .sar file will include the path prefix
+specified in the -srcar switch. This means the .sar file cannot be successfully moved
+to another location.
+
+For Phobos, std.sar is approximately 11 megabytes in size.
+
+To use the std.sar file, nothing needs to be changed in the user's build system.
+dmd will automatically prefer using any .sar file it finds. To disable using
+.sar files, which would be necessary when doing development of the source files,
+select one of the following:
+
+1. delete the .sar file
+
+2. use the -srcar=off compiler switch. -srcar=on turns it on, and is the default
+setting
+
+Trying out .sar with simple programs like hello world yield a negligible difference
+in compile speed. It's unlikely a larger program would show any particular trend
+in performance.
+
+A standalone archiver program can be easily created from the implementation in DMD.
+
+## Rationale
+
+1. All the source files in a project or library can be represented as a single file,
+making it easy to deal with.
+
+2. To compile all the source files at once with DMD, the command line can get
+extremely long, and certainly unwieldy. With .sar files, you may not even need
+a makefile or builder, just:
+
+dmd project.sar
+
+3. In Phobos (and most code), the tendency is to lump a lot of only marginally related
+functions into one file. This is likely because of the inconvenience of multiple files.
+In std.algorithm, for example, the individual algorithms could be placed into multiple
+files, since they don't refer to each other. This could also help the people who
+don't want the automatic "friend" status of declarations within a single module.
+.sar files can make much more granular modules more attractive.
```

Comment:
.sar is only a distribution file. When developing code, you're not hex editing the .sar file, you're dealing with the separate small files all the same, so it's not more convenient or attractive.

Reply:
I never edit the files in /usr/include or any utilities I download and compile from source. I only edit the project I am actively working on. I doubt many users edit dub projects other than the maintainers.

Reply:
I constantly jump in the editor to Phobos functions, to check unittests or the implementation. The same for source code of dependant libraries. I don't see how having plain source code hidden in a custom archive is more convenient while developing.

Reply:
This is not a zip, it's a *custom* format, not supported by any editor out there. So of course we need to extract it every time.

Reply:
Of course this needs to be implemented in the plugin level. And this is a very simple format that even my own engine has it implemented too.

Reply:
Yeah I wouldn't overthink it. IDE's have the facilities to support it.

We would need to make sure the big D plugins are updated to use it before we ship anything with it.

Reply:
Dustmite is another one that comes to mind. @CyberShadow